File: leetcode/230. Kth Smallest Element in a BST/230.cpp
// https://helloacm.com/how-to-find-the-kth-smallest-element-in-a-bst-tree-using-java-c/
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> list;
        dfs(root, list);
        return list[k - 1];
    }

private:
    void dfs(TreeNode* root, vector<int> &list) {
        if (root == NULL) return;
        dfs(root->left, list);
        list.push_back(root->val);
        dfs(root->right, list);
    }
};

File: leetcode/230. Kth Smallest Element in a BST/230.py
# https://helloacm.com/teaching-kids-programming-kth-smallest-element-in-a-bst-via-recursive-inorder-traversal-algorithm/
# https://leetcode.com/problems/kth-smallest-element-in-a-bst/
# MEDIUM, RECURSION

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:        
        def dfs(root):
            if not root:
                return
            yield from dfs(root.left)
            yield root.val
            yield from dfs(root.right)            
        return next(c for i, c in enumerate(dfs(root)) if i == k - 1)
